### DISCUSSION TITLE
Bandwidth difference between HTTPS and SSH cloning

### DISCUSSION THREAD

**Discussion by maroneze** [community/community#22844] [New to GitHub]
<p>I’d like to know if there are different bandwidth limits concerning cloning repositories via HTTPS or SSH, so that I can inform the users of my open-source repositories (this is important because one of the external tools which does the cloning of my repositories does not properly inform the user, so a slow git clone seems like a freeze, and users end up quitting because of that).</p>
<p>I noticed that when I clone one of my repositories using a <code>https://</code> URL, it seems capped to 500 KB/s (with a very stable upper limit), but if I use <code>git://</code> (SSH), the download speed is much more variable and can exceed 5 MB/s. The repository itself is fairly large (~80 MB), so this makes a lot of difference for our users (seconds instead of minutes).</p>
<p>We currently recommend users to stick with HTTPS, as suggested by Github and because it is simpler (no SSH key required), but nowhere in the documentation is this bandwidth difference mentioned, so I cannot find out if this difference is intentional and consistent, or if it is a specific situation that will not be experienced by my users.</p>
<p>Note: I went to <a href="https://github-debug.com/" rel="nofollow noopener">https://github-debug.com/</a> and tried cloning both repositories, but they are so small that the speed does not seem representative; there was no significant difference between cloning them (545 KB/s for the HTTPS one and 600 KB/s for the SSH one).</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>Please contact <a href="https://github.com/contact" rel="nofollow noopener">private support</a> so they can help investigate this issue. Download speeds are highly variable depending on hardware, software, location, and the route over the Internet that the data takes. There is rarely a one-size-fits-all answer but private support can help investigate if there is a problem.</p>

**Comment by maroneze** [comment]
<p>When I posted the question, I assumed this was a standard policy written somewhere, and not a specific issue, since both protocols work anyway.</p>
<p>I’ll consider private support, but since this seems to concern few people, I’m assuming it must be specific to my case and not worthy of much attention.</p>

**Reply by WorldDogs** [reply]
2023 same question, ssh is much slower than https on my conputer

**Comment by xiaoronglv** [comment]
<p>This issue happened to me too. ssh is extremely slow, I have to replace the origin address from <code>git@github.com:xxxxx</code> to <code>http:xxxxx</code></p>

**Comment by wangrongding** [comment]
Waiting for a good answer. 👀 👀

**Comment by xmedeko** [comment]
According to my experiments, HTTPS is faster for tiny data transfers, e.g. `git fetch` or `git push` with no or small data payloads.

**Comment by xmedeko** [comment]
See also this SO answer https://stackoverflow.com/a/50567020/254109 : GitHub does CPU throttling and the difference may be caused by authenticated vs unauthenticated access.

If you have a private GitHub repo, then you have to use authenticated HTTPS access. However, I do not know how to enforce authenticated HTTPS access for cloning public repo.

**Reply by ulidtko** [reply]
I'm getting fetch speeds **~10 KiB/s** — both on HTTPS & SSH urls (the ssh ones, ofcourse, being authenticated).

At the same time, speedtests rate my uplink speed at 100-150 Mbps — so there's definitely something on GH side. However I don't believe it's auth-related.

**Reply by Abdelalihub** [reply]
Similar issue here too
**Receiving objects:   0% (9/4533), 1.30 MiB | 102.00 KiB/s**

**Comment by GJCav** [comment]
I am experiencing the same issue but in my situation, HTTPS connection is faster. For example, when cloning via HTTPS, I can fully utilize my bandwidth, achieving speeds over 20MB/s. However, SSH cloning tops out at approximately 5MB/s.

Additional details:

I am based in mainland China, where, as is widely known, the Great Firewall restricts nearly all GitHub connections, be they HTTPS or SSH. To circumvent these restrictions, I employ a proxy for both connection types. Below are my proxy configurations:


SSH Configuration (~/.ssh/config):

``` 
Host github.com
  Port 443
  HostName ssh.github.com
  User git

# my proxy serves on 127.0.0.1:7890
  ProxyCommand "D:\programs\Git\mingw64\bin\connect.exe" -H localhost:7890 %h %p
```

Git Configuration (.gitconfig):

```
[http]
	proxy = http://127.0.0.1:7890
[https]
	proxy = http://127.0.0.1:7890
```

To evaluate the connection speeds, I cloned the NASA-3D-Resources repository (https://github.com/nasa/NASA-3D-Resources), which is relatively large. Here are the results:

* over SSH
    * speed: ~ 5MB/s
    * command: `git clone git@github.com:nasa/NASA-3D-Resources.git`
* over HTTPS
    * speed: > 20MB/s
    * command: `git clone https://github.com/nasa/NASA-3D-Resources.git`

Similarly, when pushing changes to one of my private repositories, I noted the same disparity in speed:

* over SSH
    * speed: ~ 2MB/s
* over HTTPS
    * speed: > 10MB/s

My colleagues have replicated these tests and obtained similar results. Could anyone help clarify why there is such a difference in speed between SSH and HTTPS connections under these settings?

Thank you for any insights or suggestions you may offer.

**Reply by ulidtko** [reply]
Hi, let me suggest another metric you can measure: **packet loss** to github.com.

I'd just hit this issue again; could not clone a tiny 3MiB repo repeatedly. Turned out, my VPN gateway is giving consistent 70% packet loss to github.com today:

![Screenshot_20240427_120802](https://github.com/community/community/assets/365338/706c24f7-517c-4a7e-a49d-3b388a373101)
![Screenshot_20240427_121117](https://github.com/community/community/assets/365338/56c3b631-99c0-4285-b542-8f9be2627f88)

Swithing to another VPN endpoint gave me 0% packet loss, and instant repo clone (at ~tens of MiB/s).

I understand that your setup is different (SS proxy, not VPN), but it could have a similar packet loss issue with strong influence on speed. It could also very well be dependent on port number / traffic encapsulation format (OpenSSH vs TLS).
